`TestManagedRegistry_externalLabels` is flaky

The `TestManagedRegistry_externalLabels` test is flaky, and will randomly fail.

```sh
➜  tempo git:(main) go test ./modules/generator/registry --run TestManagedRegistry_externalLabels -v --count=1
=== RUN   TestManagedRegistry_externalLabels
--- PASS: TestManagedRegistry_externalLabels (0.00s)
PASS
ok  	github.com/grafana/tempo/modules/generator/registry	0.338s
➜  tempo git:(main) go test ./modules/generator/registry --run TestManagedRegistry_externalLabels -v --count=1
=== RUN   TestManagedRegistry_externalLabels
    registry_test.go:252: 
        	Error Trace:	/Users/lasse.hels/Projects/tempo/modules/generator/registry/registry_test.go:252
        	            				/Users/lasse.hels/Projects/tempo/modules/generator/registry/registry_test.go:150
        	Error:      	elements differ
        	            	
        	            	extra elements in list A:
        	            	([]interface {}) (len=2) {
        	            	 (registry.sample) {
        	            	  l: (labels.Labels) (len=3) {
        	            	   (labels.Label) {
        	            	    Name: (string) (len=5) "__foo",
        	            	    Value: (string) (len=3) "bar"
        	            	   },
        	            	   (labels.Label) {
        	            	    Name: (string) (len=22) "__metrics_gen_instance",
        	            	    Value: (string) (len=13) "MMDKF7QLWWFW0"
        	            	   },
        	            	   (labels.Label) {
        	            	    Name: (string) (len=8) "__name__",
        	            	    Value: (string) (len=10) "my_counter"
        	            	   }
        	            	  },
        	            	  t: (int64) 1693557980698,
        	            	  v: (float64) 0
        	            	 },
        	            	 (registry.sample) {
        	            	  l: (labels.Labels) (len=3) {
        	            	   (labels.Label) {
        	            	    Name: (string) (len=5) "__foo",
        	            	    Value: (string) (len=3) "bar"
        	            	   },
        	            	   (labels.Label) {
        	            	    Name: (string) (len=22) "__metrics_gen_instance",
        	            	    Value: (string) (len=13) "MMDKF7QLWWFW0"
        	            	   },
        	            	   (labels.Label) {
        	            	    Name: (string) (len=8) "__name__",
        	            	    Value: (string) (len=10) "my_counter"
        	            	   }
        	            	  },
        	            	  t: (int64) 1693557980698,
        	            	  v: (float64) 1
        	            	 }
        	            	}
        	            	
        	            	
        	            	extra elements in list B:
        	            	([]interface {}) (len=2) {
        	            	 (registry.sample) {
        	            	  l: (labels.Labels) (len=3) {
        	            	   (labels.Label) {
        	            	    Name: (string) (len=22) "__metrics_gen_instance",
        	            	    Value: (string) (len=13) "MMDKF7QLWWFW0"
        	            	   },
        	            	   (labels.Label) {
        	            	    Name: (string) (len=5) "__foo",
        	            	    Value: (string) (len=3) "bar"
        	            	   },
        	            	   (labels.Label) {
        	            	    Name: (string) (len=8) "__name__",
        	            	    Value: (string) (len=10) "my_counter"
        	            	   }
        	            	  },
        	            	  t: (int64) 1693557980698,
        	            	  v: (float64) 0
        	            	 },
        	            	 (registry.sample) {
        	            	  l: (labels.Labels) (len=3) {
        	            	   (labels.Label) {
        	            	    Name: (string) (len=22) "__metrics_gen_instance",
        	            	    Value: (string) (len=13) "MMDKF7QLWWFW0"
        	            	   },
        	            	   (labels.Label) {
        	            	    Name: (string) (len=5) "__foo",
        	            	    Value: (string) (len=3) "bar"
        	            	   },
        	            	   (labels.Label) {
        	            	    Name: (string) (len=8) "__name__",
        	            	    Value: (string) (len=10) "my_counter"
        	            	   }
        	            	  },
        	            	  t: (int64) 1693557980698,
        	            	  v: (float64) 1
        	            	 }
        	            	}
        	            	
        	            	
        	            	listA:
        	            	([]registry.sample) (len=2) {
        	            	 (registry.sample) {
        	            	  l: (labels.Labels) (len=3) {
        	            	   (labels.Label) {
        	            	    Name: (string) (len=5) "__foo",
        	            	    Value: (string) (len=3) "bar"
        	            	   },
        	            	   (labels.Label) {
        	            	    Name: (string) (len=22) "__metrics_gen_instance",
        	            	    Value: (string) (len=13) "MMDKF7QLWWFW0"
        	            	   },
        	            	   (labels.Label) {
        	            	    Name: (string) (len=8) "__name__",
        	            	    Value: (string) (len=10) "my_counter"
        	            	   }
        	            	  },
        	            	  t: (int64) 1693557980698,
        	            	  v: (float64) 0
        	            	 },
        	            	 (registry.sample) {
        	            	  l: (labels.Labels) (len=3) {
        	            	   (labels.Label) {
        	            	    Name: (string) (len=5) "__foo",
        	            	    Value: (string) (len=3) "bar"
        	            	   },
        	            	   (labels.Label) {
        	            	    Name: (string) (len=22) "__metrics_gen_instance",
        	            	    Value: (string) (len=13) "MMDKF7QLWWFW0"
        	            	   },
        	            	   (labels.Label) {
        	            	    Name: (string) (len=8) "__name__",
        	            	    Value: (string) (len=10) "my_counter"
        	            	   }
        	            	  },
        	            	  t: (int64) 1693557980698,
        	            	  v: (float64) 1
        	            	 }
        	            	}
        	            	
        	            	
        	            	listB:
        	            	([]registry.sample) (len=2) {
        	            	 (registry.sample) {
        	            	  l: (labels.Labels) (len=3) {
        	            	   (labels.Label) {
        	            	    Name: (string) (len=22) "__metrics_gen_instance",
        	            	    Value: (string) (len=13) "MMDKF7QLWWFW0"
        	            	   },
        	            	   (labels.Label) {
        	            	    Name: (string) (len=5) "__foo",
        	            	    Value: (string) (len=3) "bar"
        	            	   },
        	            	   (labels.Label) {
        	            	    Name: (string) (len=8) "__name__",
        	            	    Value: (string) (len=10) "my_counter"
        	            	   }
        	            	  },
        	            	  t: (int64) 1693557980698,
        	            	  v: (float64) 0
        	            	 },
        	            	 (registry.sample) {
        	            	  l: (labels.Labels) (len=3) {
        	            	   (labels.Label) {
        	            	    Name: (string) (len=22) "__metrics_gen_instance",
        	            	    Value: (string) (len=13) "MMDKF7QLWWFW0"
        	            	   },
        	            	   (labels.Label) {
        	            	    Name: (string) (len=5) "__foo",
        	            	    Value: (string) (len=3) "bar"
        	            	   },
        	            	   (labels.Label) {
        	            	    Name: (string) (len=8) "__name__",
        	            	    Value: (string) (len=10) "my_counter"
        	            	   }
        	            	  },
        	            	  t: (int64) 1693557980698,
        	            	  v: (float64) 1
        	            	 }
        	            	}
        	Test:       	TestManagedRegistry_externalLabels
--- FAIL: TestManagedRegistry_externalLabels (0.00s)
FAIL
FAIL	github.com/grafana/tempo/modules/generator/registry	0.346s
FAIL
➜  tempo git:(main) go test ./modules/generator/registry --run TestManagedRegistry_externalLabels -v --count=1
=== RUN   TestManagedRegistry_externalLabels
--- PASS: TestManagedRegistry_externalLabels (0.00s)
PASS
ok  	github.com/grafana/tempo/modules/generator/registry	0.346s
```

Looks like the label ordering might be the root cause. Other tests _may_ also be affected, as the `collectRegistryMetricsAndAssert` function is used in several tests in `registry_test.go`.

Tested on commit https://github.com/grafana/tempo/commit/13bbd124f55e8db33cdef2f2ab510edaaebb743a.

--- a/modules/generator/registry/registry_test.go
+++ b/modules/generator/registry/registry_test.go
@@ -3,7 +3,6 @@ package registry
 import (
 	"context"
 	"fmt"
-	"github.com/stretchr/testify/require"
 	"math/rand"
 	"os"
 	"sort"
@@ -12,6 +11,7 @@ import (
 
 	"github.com/go-kit/log"
 	"github.com/stretchr/testify/assert"
+	"github.com/stretchr/testify/require"
 )
 
 func TestManagedRegistry_concurrency(*testing.T) {

--- a/modules/generator/registry/registry_test.go
+++ b/modules/generator/registry/registry_test.go
@@ -2,6 +2,7 @@ package registry
 
 import (
 	"context"
+	"fmt"
 	"math/rand"
 	"os"
 	"testing"
@@ -9,6 +10,7 @@ import (
 
 	"github.com/go-kit/log"
 	"github.com/stretchr/testify/assert"
+	"github.com/stretchr/testify/require"
 )
 
 func TestManagedRegistry_concurrency(*testing.T) {
@@ -249,7 +251,53 @@ func collectRegistryMetricsAndAssert(t *testing.T, r *ManagedRegistry, appender 
 
 	assert.Equal(t, true, appender.isCommitted)
 	assert.Equal(t, false, appender.isRolledback)
-	assert.ElementsMatch(t, expectedSamples, appender.samples)
+
+	actualSamples := appender.samples
+
+	require.Equal(t, len(expectedSamples), len(actualSamples))
+
+	// This loop performs equality comparison of expected and actual samples agnostic of order of both samples
+	// and labels within samples.
+	for _, expected := range expectedSamples {
+		if i, ok := match(expected, actualSamples); ok {
+			// Remove matching sample to prevent multiple matches on the same sample.
+			actualSamples = remove(actualSamples, i)
+			continue
+		}
+
+		t.Fatalf("no match found for expected sample %s", expected.String())
+	}
+}
+
+// match checks if actualSamples contains at least one sample that matches the expected sample.
+// The returned integer is the index of the found match; the returned boolean indicates whether a match was found.
+func match(expected sample, actualSamples []sample) (int, bool) {
+	for i, actual := range actualSamples {
+		if expected.t != actual.t {
+			continue
+		}
+
+		if expected.v != actual.v {
+			continue
+		}
+
+		// Rely on the fact that Go prints map keys in sorted order.
+		// See https://tip.golang.org/doc/go1.12#fmt.
+		if fmt.Sprint(expected.l.Map()) != fmt.Sprint(actual.l.Map()) {
+			continue
+		}
+
+		return i, true
+	}
+
+	return 0, false
+}
+
+// remove item from slice.
+// https://stackoverflow.com/a/37335777
+func remove(samples []sample, i int) []sample {
+	samples[i] = samples[len(samples)-1]
+	return samples[:len(samples)-1]
 }
 
 type mockOverrides struct {

--- a/modules/generator/registry/registry_test.go
+++ b/modules/generator/registry/registry_test.go
@@ -3,14 +3,15 @@ package registry
 import (
 	"context"
 	"fmt"
+	"github.com/stretchr/testify/require"
 	"math/rand"
 	"os"
+	"sort"
 	"testing"
 	"time"
 
 	"github.com/go-kit/log"
 	"github.com/stretchr/testify/assert"
-	"github.com/stretchr/testify/require"
 )
 
 func TestManagedRegistry_concurrency(*testing.T) {
@@ -253,51 +254,28 @@ func collectRegistryMetricsAndAssert(t *testing.T, r *ManagedRegistry, appender 
 	assert.Equal(t, false, appender.isRolledback)
 
 	actualSamples := appender.samples
-
 	require.Equal(t, len(expectedSamples), len(actualSamples))
 
-	// This loop performs equality comparison of expected and actual samples agnostic of order of both samples
-	// and labels within samples.
-	for _, expected := range expectedSamples {
-		if i, ok := match(expected, actualSamples); ok {
-			// Remove matching sample to prevent multiple matches on the same sample.
-			actualSamples = remove(actualSamples, i)
-			continue
-		}
+	// Ensure that both slices are ordered consistently.
+	for _, slice := range [][]sample{expectedSamples, actualSamples} {
+		sort.Slice(slice, func(i, j int) bool {
+			this := slice[i]
+			next := slice[j]
 
-		t.Fatalf("no match found for expected sample %s", expected.String())
+			// The actual order doesn't matter, the only thing that matters is that it is consistent.
+			return this.String() < next.String()
+		})
 	}
-}
 
-// match checks if actualSamples contains at least one sample that matches the expected sample.
-// The returned integer is the index of the found match; the returned boolean indicates whether a match was found.
-func match(expected sample, actualSamples []sample) (int, bool) {
-	for i, actual := range actualSamples {
-		if expected.t != actual.t {
-			continue
-		}
+	for i, expected := range expectedSamples {
+		actual := actualSamples[i]
 
-		if expected.v != actual.v {
-			continue
-		}
-
+		assert.Equal(t, expected.t, actual.t)
+		assert.Equal(t, expected.v, actual.v)
 		// Rely on the fact that Go prints map keys in sorted order.
 		// See https://tip.golang.org/doc/go1.12#fmt.
-		if fmt.Sprint(expected.l.Map()) != fmt.Sprint(actual.l.Map()) {
-			continue
-		}
-
-		return i, true
+		assert.Equal(t, fmt.Sprint(expected.l.Map()), fmt.Sprint(actual.l.Map()))
 	}
-
-	return 0, false
-}
-
-// remove item from slice.
-// https://stackoverflow.com/a/37335777
-func remove(samples []sample, i int) []sample {
-	samples[i] = samples[len(samples)-1]
-	return samples[:len(samples)-1]
 }
 
 type mockOverrides struct {

